### PR TITLE
fix(training/lerobot): embodiment and tune reach the policy config that declares them

### PR DESCRIPTION
**Pull request by cagataycali**
`LerobotTrainer` lists `groot` among the policy types it trains, and lerobot's own `GrootConfig` declares an `embodiment_tag` plus four per-component switches. Two `TrainSpec` fields are documented as exactly those requests, and the trainer read neither.

what a GR00T tuning request became

Every cell is measured, both trees driven in separate subprocesses with a provenance assert (`("policy_cfg.embodiment_tag = spec.embodiment" in getsource(lerobot)) == (label == "after")`).

## The headline: the request came back inverted on all four components

`spec.embodiment` was read **nowhere** in the module and `spec.tune` in exactly one place - the lora/expert-only mutual-exclusion check. So a spec asking to unfreeze the language backbone and vision tower and leave the projector and action head alone built the opposite config, against the default head, and `validate()` returned `[]`:

```
>>> spec = TrainSpec(embodiment="so101_follower",
...                  tune={"llm": True, "visual": True, "projector": False, "diffusion": False}, ...)
>>> LerobotTrainer(policy_type="groot").validate(spec)
[]
>>> p = LerobotTrainer(policy_type="groot").build_config(spec).policy   # GrootConfig
>>> p.tune_llm, p.tune_visual, p.tune_projector, p.tune_diffusion_model
(False, False, True, True)          # asked for (True, True, False, False)
>>> p.embodiment_tag
'new_embodiment'                    # asked for 'so101_follower'
```

Four toggles, four inversions, plus a projector head the caller did not name. A GR00T fine-tune is *specified* by that dict, so the run trains a different recipe end to end and nothing in the launch says so - an unforwarded toggle is indistinguishable from never having asked.

## The rule was already in the same method, twice

`_build_policy_config` handles two other `TrainSpec` fields that name a policy-config field, and both do the right thing 20 lines away:

| field | applied when the config declares it | refused when it does not |
|---|---|---|
| `learning_rate` -> `optimizer_lr` | yes | `ValueError` naming the field |
| `relative_actions` -> `use_relative_actions` | yes | `ValueError` + `validate()` problem |
| `method="expert_only"` -> `train_expert_only` | yes | `validate()` problem |
| **`embodiment` -> `embodiment_tag`** | **no** | **no** |
| **`tune` -> `tune_llm` / ...** | **no** | **no** |

And `_policy_supports_expert_only`'s own docstring states the harm for its field, which is the harm here verbatim: requesting it on a policy whose config lacks the field "is either a silent no-op (in-process `build_config` guards on `hasattr` and never flips the flag, so the run silently full-finetunes the backbone while reporting success) ... so it must be gated by the policy's actual capability".

## The change

Both fields now reach the config, and a policy whose config lacks them refuses instead of dropping the request. Capability is probed off lerobot's **registered config class** (`_policy_supports_embodiment_tag`, `_policy_tune_components`) - the same dataclass-field lookup its three sibling probes use, with the documented offline fallback - so it is not a hand-kept list:

| lerobot policy type | `embodiment_tag` | `tune_*` |
|---|---|---|
| `groot` | yes | 4 |
| the other 18 (`act`, `pi0`, `smolvla`, `molmoact2`, `xvla`, ...) | - | - |

Measured over every type `_lerobot_policy_types()` returns against lerobot 0.6.2, and pinned, so a policy lerobot adds with either surface is honoured on arrival rather than needing an edit here.

Two further ways a toggle went quiet, both closed:

- **a key naming no component.** `tune={"vision": True}` (lerobot spells it `visual`) matched nothing to forward and trained the default. Now refused, naming the accepted set. `expert_only` stays accepted - it is a `method`, and the mutual-exclusion check's input.
- **`build_command`.** It is documented as the argv the typed config maps to, so it carries `--policy.embodiment_tag` and `--policy.tune_*`, iterated in a canonical order rather than the caller's dict order.

`TrainSpec.embodiment`'s docstring said "LeRobot infers it from dataset features", which stopped being true when lerobot shipped `GrootConfig`; corrected, along with `training/groot.py`'s "the model-tuning surface lerobot does NOT have" comment beside the identical four flags.

## Tests

`tests/training/test_embodiment_and_tune_reach_the_policy_config.py`, 26 cells. On `upstream/main` with only the two capability probes copied in as the instrument and the trainer's read sites untouched: **17 failed / 9 passed**. After: 26 passed.

Mutants, each measured over those 26 cells (harness restores the tree and re-reports 0 as shipped):

| mutation | cells failing |
|---|---|
| the component toggles are not applied (pre-fix) | 6 |
| `validate()` reports neither request (pre-fix) | 6 |
| a key naming no component is accepted | 4 |
| the capability probe answers yes for every policy | 3 |
| `embodiment` is not applied to the config (pre-fix) | 1 |
| `expert_only` is swept as a component | 1 |
| the toggles follow the caller's dict order | 1 |
| the documented offline fallback drifts (`groot` dropped) | 1 |

The last row is why the derivation is pinned rather than the values: the fallback set is asserted to *be* what the live probe says, so a lerobot that stops registering `groot` cannot leave the offline answer lying.

## Gate

| check | result |
|---|---|
| `pytest tests/` | **50876 passed / 307 skipped / 0 failed**, 29m11s |
| `ruff check` + `format --check` (`strands_robots tests tests_integ`) | clean, 1946 files |
| `mypy` | Success, 1897 source files |

## LOC

| path | delta |
|---|---|
| `strands_robots/training/lerobot.py` | +154 |
| `strands_robots/training/base.py` | +15 / -2 |
| `strands_robots/tools/train_policy.py` | +12 / -2 |
| `strands_robots/training/groot.py` | +5 / -1 |
| `docs/training/overview.md` | +19 / -1 |
| tests | +205 |

Net positive. Of the 154 source lines the two derived probes and their documented fallbacks are the bulk (the shape three siblings in the same module already use), then the two `validate()` reporters, then the apply/refuse sites. No new public API - the two fields were already declared, documented and reachable through the `train_policy` tool.

### Round 1 (`ddcb3be9d`)

One concern: `_tune_component_fields` coerced each toggle with `bool(...)`, so a truthy spelling of off (`"false"`, `"no"`, `"off"`, `"0"`) trained the component the caller asked to freeze, with `validate()` clean - the headline inversion reintroduced on the value axis. Each recognized component's value now goes through `boolean_flag_error` under its own name (`tune['llm']`), the checked value is forwarded uncoerced, and a builder reached without `validate()` raises the same refusal as a `ValueError`. One owner, `_tune_value_error`, for all three surfaces. Pin: `TestATuneValueIsCheckedNotRead`, 22 cells, 14 fail on the coerced read. Fragment and `docs/training/overview.md` each gain one sentence on the value axis.

### Round 2 (`30b88b93`)

The required check went red on round 1's head in the lint step alone: `mypy` inferred `list[str]` and `list[str | int | None]` for the two module-level spelling lists in the new test class and refused their `+` in the parametrize (one error, 1946 files). Both are now annotated `list[object]`; no behaviour changes. Tests, `ruff`, `mypy` clean on the touched file.

### CODE REVIEW

File: strands_robots/training/lerobot.py [hunk 1]
```diff
@@ -1083,9 +1150,73 @@ def _validate_policy(self, spec: TrainSpec) -> list[str]:
...
+    def _tune_component_fields(self, spec: TrainSpec) -> dict[str, bool]:
+        """The requested component toggles, keyed by lerobot policy-config field.
+
+        Iterates :data:`_TUNE_COMPONENT_FIELDS` rather than ``spec.tune`` so the
+        order is the canonical one whatever order the caller's dict has - which
+        is what makes :meth:`build_command`'s argv comparable run to run.
+        ``expert_only`` is excluded: it is a ``method``, not a component.
+        """
+        return {
+            field: bool(spec.tune[component])
```
Comment (yinsong1986) [inline strands_robots/training/lerobot.py]: [MUST FIX] `field: bool(spec.tune[component])` reads a posture flag by truthiness. Every non-empty string is truthy, so `tune={"llm": "false"}` (or `"no"`, `"off"`, `"0"` - the spellings a YAML/JSON-sourced config actually carries) passes the new `_tune_component_problems` clean (it checks keys only), builds `tune_llm=True`, and the run silently trains the language backbone the caller asked to freeze - the exact inverted-recipe failure this PR's description names as the headline bug, reintroduced on the value axis instead of the key axis. Nothing raises and nothing logs, so the wrong recipe is indistinguishable from the right one.

This is the pinned convention in AGENTS.md > "Posture flags are checked, never read by truthiness" (`boolean_flag_error` is the shared domain; the numeric domains reject `bool` precisely so this one can require it), and this same module already follows it for its other boolean spec fields: `resume` and `streaming` are preflighted through `_validate._posture_flag_problems` -> `boolean_flag_error`, with the comment at `_validate_policy`'s streaming capture stating the rationale. The four tune values select which model components train, so they are squarely in the flag class.

Fix: in `_tune_component_problems`, run each recognized component's value through `strands_robots.utils.boolean_flag_error(spec.tune[c], f"tune['{c}']", ...)` so `validate()` refuses the misread posture by the flag's own name, and drop the `bool(...)` coercion here (forward the checked value). Must-fix because the failure is silent wrong training on the surface this PR ships as fixed: a `validate() == []` spec trains the inverse of the request, which is the data-integrity class (silent wrong behaviour reported as success) rather than a coverage or style gap.

**Review by yinsong1986 — COMMENTED** [review]
## Summary

Wires `TrainSpec.embodiment` and `TrainSpec.tune` through to the lerobot policy config that declares them (`GrootConfig.embodiment_tag` / `tune_llm` / `tune_visual` / `tune_projector` / `tune_diffusion_model`), and applies the same apply-or-refuse rule `learning_rate` and `relative_actions` already follow in `_build_policy_config`: a policy whose config lacks the field gets a `validate()` problem plus a `ValueError` at build, instead of a `hasattr`-guarded silent drop that trains the config defaults while reporting success. Capability is probed off lerobot's registered config classes via the same dataclass-field lookup the three sibling probes use, with the offline fallback pinned against the live answer so it cannot drift. `build_command` carries the matching `--policy.*` flags in a canonical order, and a `tune` key naming no component is refused by name. One must-fix inline: the tune toggle *values* are read by truthiness, which reintroduces the same silent-inversion class this PR closes, one axis over.

## What's good

- The capability probes are derived from lerobot's registry rather than hand-kept lists, and `TestTheCapabilityProbesTrackLerobot` pins both the live agreement and the offline fallback, so a lerobot that adds or drops the surface is graded on arrival.
- Refusal messages name the supporting policy set and the remedy, matching the `relative_actions`/`expert_only` siblings verbatim.
- Changelog fragment named for the PR number, docs (`train_policy` docstring, `TrainSpec` docstring, `docs/training/overview.md`, the stale `groot.py` comment) corrected in the same PR.

File: strands_robots/training/lerobot.py (hunk 1 above)
Comment (cagataycali) [reply strands_robots/training/lerobot.py]: Fixed in `ddcb3be9d`, taking the suggested shape.

`_tune_component_problems` now grades every recognized component's value through `boolean_flag_error(spec.tune[c], f"tune['{c}']", self.provider_name)`, so `validate()` refuses the misread posture under the flag's own name; `_tune_component_fields` forwards the checked value and the `bool(...)` coercion is gone. The domain has one owner (`_tune_value_error`) that both the reporter and the two builders read, so a builder reached without `validate()` raises the same text as a `ValueError` - the shape `_build_policy_config` already uses for a component the policy lacks.

Measured on the new head, `tune={"llm": v}` with `policy_type="groot"`:

| `v` | `validate()` | `build_command` |
|---|---|---|
| `"false"` / `"no"` / `"off"` / `"0"` | `lerobot_local: tune['llm'] must be a boolean, got 'false'. ...` | `ValueError`, same text |
| `1` / `0` / `None` / `"true"` | refused, same shape | `ValueError` |
| `True` / `False` / `np.bool_(False)` | no `tune['llm']` problem | `--policy.tune_llm=true` / `false` as given |

Pin: `TestATuneValueIsCheckedNotRead` in the same test file, 22 cells - the four truthy spellings of off on all three surfaces, the rest of the non-boolean domain, all four components graded (not only the first), the unknown-key refusal surviving alongside the value refusal, and the boolean controls. Reverting to the coerced read fails 14 of them.

Fragment and `docs/training/overview.md` each gain one sentence stating the value axis, so neither describes a key-only refusal.

---

> This response was generated by an AI contributor. Verified with `ruff check`/`ruff format --check`/`mypy` clean on the touched files and the test file run against both the fixed and the mutated tree.

**Review by yinsong1986 — APPROVED** [review]
## Summary

Wires `TrainSpec.embodiment` and `TrainSpec.tune` through to the lerobot policy config that declares them (`GrootConfig.embodiment_tag` / `tune_llm` / `tune_visual` / `tune_projector` / `tune_diffusion_model`), applying the same apply-or-refuse rule `learning_rate` and `relative_actions` already follow in `_build_policy_config`: a policy whose config lacks the field gets a `validate()` problem and a `ValueError` at build, instead of the pre-fix silent drop that trained the default head and the inverse of all four component toggles while reporting success. Capability is probed off lerobot's registered config class (the same dataclass-field lookup the three sibling probes use) with documented offline fallbacks that are pinned to equal the live answer, so a lerobot policy that grows either surface is honoured without an edit here. A `tune` key naming no component is refused by name, `build_command`'s argv carries the same flags in canonical order, and each toggle's value is checked through `boolean_flag_error` under its own name (`tune['llm']`) with one owner (`_tune_value_error`) shared by `validate()` and both builders - resolving the round-1 truthiness finding. Verified at head that `build_command` is the documented pure argv-parity helper rather than a launch path, so the `embodiment` interpolation there is not a subprocess surface; the new files carry no non-ASCII, no host paths, and the changelog fragment, tool `Args:` entries and docs all moved in the same change.

## What's good

- The refusal shape, the capability probes and the offline-fallback pinning all mirror the module's existing `_policy_supports_expert_only` / `_policy_supports_relative_actions` siblings rather than inventing a second pattern.
- Value checking routes through the shared `boolean_flag_error` domain per AGENTS.md > "Posture flags are checked, never read by truthiness", with `TestATuneValueIsCheckedNotRead` parametrized over the refused spellings and a numpy-bool control.
- `TestTheCapabilityProbesTrackLerobot` asserts the probes against the installed registry and the fallback sets against the live answer, so neither can drift silently.
